src/execute-afm.js

```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2007-2014, GoodData(R) Corporation. All rights reserved.
+import { get, mapValues, clone } from 'lodash';
+import invariant from 'invariant';
+import { ajax, post, parseJSON } from './xhr';
+import { queryString } from './util';
+
+const PAGE_SIZE = 500;
+
+function getDimensionality(execution) {
+    return get(execution, 'execution.resultSpec.dimensions', [1, 2]).length;
+}
+
+function getLimit(offset) {
+    return Array(offset.length).fill(PAGE_SIZE);
+}
+
+function fetchExecutionResult(pollingUri, offset) {
+    const query = { limit: getLimit(offset), offset };
+    const uri = pollingUri + queryString(mapValues(query, arr => arr.join(',')));
+    return ajax(uri, { method: 'GET' }).then((r) => {
+        if (r.status === 204) {
+            const err = new Error('Loading executeAfm failed: 204 No Content');
+            err.response = r;
+            throw err;
+        }
+        return r.json();
+    });
+}
+
+
+export function mergePageData(resultSoFar, { offset, data }) {
+    const rowOffset = offset[0];
+    if (resultSoFar.data[rowOffset]) { // appending to existing rows
+        for (let i = 0; i < data.length; i += 1) {
+            resultSoFar.data[i + rowOffset].push(...data[i]);
+        }
+    } else {
+        resultSoFar.data.push(...data);
+    }
+    return resultSoFar;
+}
```

Comment:
This works only for one or two dimensions. Is it ok to leave `invariant` only in the "main" function?